lib/wca_live_web/controllers/competition_controller.ex

```diff
@@ -174,4 +190,76 @@ defmodule WcaLiveWeb.CompetitionController do
         {:error, "the provided token is not valid"}
     end
   end
+
+  defp fetch_competition_by_id_or_wca_id(id_or_wca_id) do
+    case Integer.parse(id_or_wca_id) do
+      {id, _} -> Competitions.fetch_competition(id)
+      _ -> Competitions.fetch_competition_by_wca_id(id_or_wca_id)
+    end
+  end
+
+  defp format_results(competition) do
+    competition
+    |> Repo.preload(
+      competition_events: [rounds: [:competition_event, results: [:person]]],
+      people: []
+    )
+    |> competition_to_results()
+  end
+
+  defp competition_to_results(competition) do
+    result_registrant_ids =
+      for competition_event <- competition.competition_events,
+          round <- competition_event.rounds,
+          result <- round.results,
+          do: result.person.registrant_id,
+          into: MapSet.new()
+
+    %{
+      events:
+        competition.competition_events
+        |> Enum.map(&format_competition_event/1),
+      persons:
+        competition.people
+        |> Enum.filter(&(&1.registrant_id in result_registrant_ids))
+        |> Enum.map(&format_person/1)
+        |> Enum.filter(&(&1["id"] != nil))
```

Comment:
Just a detail, but each filter/map will do a separate pass on the list, we can make it into a single pass with the comprehension syntax:

```suggestion
      persons:
        for person <- competition.people, person.registrant_id in result_registrant_ids do
          format_person(person)
        end
```

Also, I don't think we need to filter by the id. I don't remember where the registrant id would be nil, but people that have results definitely do have the id.

Reply:
I did this because I'd get non-competing users like "PNW Cubing", the users we use to direct invoices to the org. For some reason they'd have nil IDs.

![image](https://github.com/user-attachments/assets/23908307-5e30-49de-8993-84f362d153e2)

But with your suggestion, they already get filtered out with `|> Enum.filter(&(&1.registrant_id in result_registrant_ids))`